### PR TITLE
Sync feature branch with main

### PR DESCRIPTION
**Pull request by chatton**
## PSA: **Do not squash this commit** unless we want to handle the conflicts a second time

A few notable changes.

~1. I needed to add a replace statement to the callbacks go.mod (I think we can figure out how to remove it before we tag anything)
it is pulled in via the testing package~

```
go: github.com/cosmos/ibc-go/modules/apps/callbacks tested by
        github.com/cosmos/ibc-go/modules/apps/callbacks.test imports
        github.com/cosmos/ibc-go/v8/testing imports
        github.com/cosmos/ibc-go/modules/light-clients/08-wasm/types: missing go.sum entry for module providing package github.com/cosmos/ibc-go/modules/light-clients/08-wasm/types (imported by github.com/cosmos/ibc-go/v8/testing); to add:
        go get github.com/cosmos/ibc-go/v8/testing@v8.0.0
```
~2. Because of the above, I needed to make our CI play nice with our callbacks simapp now requiring the wasm static libs. (again hopefully temporary)~

Any of these changes that we want to revert I think we can do a in a follow up, this PR is already quite messy and I think it would be easier to figure out these issues in isolation instead of adding to this monster of a PR 👹 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/11-structure.md) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to be used for the changelog entry in the PR description for review.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.
- [ ] Review `Codecov Report` in the comment section below once CI passes.

### CODE REVIEW

**Comment by crodriguezvega** [comment]
1328 files changed 😱

**Comment by chatton** [comment]
> 1328 files changed 😱

I definitely was not expecting to go down this rabbit hole yesterday 🥲

**Comment by crodriguezvega** [comment]
> I definitely was not expecting to go down this rabbit hole yesterday 🥲

Unfortunately there is always a 🐰🕳️ lurking around every corner... But thank so much for doing this, @chatton!

**Comment by colin-axner** [comment]
Now that the wasm dep has been removed from ibc-go go.mod, the callbacks replace can be removed?

File: .github/workflows/callbacks.yml [hunk 1]
```diff
@@ -26,21 +26,23 @@ jobs:
...
+      - uses: actions/checkout@v4
       - uses: actions/setup-go@v4
         with:
           go-version: '1.21'
       - name: Build callbacks-module
-        run: |
-          cd modules/apps/callbacks
-          GOARCH=${{ matrix.go-arch }} go build ./...
+        run: docker build . -f modules/apps/callbacks/Dockerfile --build-arg GOARCH=${{ matrix.go-arch }}
+
```
Comment (colin-axner) [inline .github/workflows/callbacks.yml:40]: ditto

File: modules/apps/callbacks/Dockerfile [hunk 2]
```diff
@@ -3,7 +3,12 @@ FROM golang:1.21-alpine3.18 as builder
 RUN set -eux; apk add --no-cache git libusb-dev linux-headers gcc musl-dev make;
 
 ENV GOPATH=""
-ENV GOMODULE="on"
+ARG GOARCH="amd64"
+# Grab the static library and copy it to location that will be found by the linker flag `-lwasmvm_muslc`.
+# TODO: nice to have: a script to download version matching the wasmvm version in go.mod.
+ADD https://github.com/CosmWasm/wasmvm/releases/download/v1.2.4/libwasmvm_muslc.x86_64.a /lib/libwasmvm_muslc.x86_64.a
+RUN sha256sum /lib/libwasmvm_muslc.x86_64.a | grep ce3d892377d2523cf563e01120cb1436f9343f80be952c93f66aa94f5737b661
+RUN cp /lib/libwasmvm_muslc.x86_64.a /lib/libwasmvm_muslc.a
```
Comment (colin-axner) [inline modules/apps/callbacks/Dockerfile:11]: should we open a tracking issue?

File: modules/apps/callbacks/go.mod [hunk 3]
```diff
@@ -2,74 +2,91 @@ module github.com/cosmos/ibc-go/modules/apps/callbacks
...
 go 1.21
 
-replace github.com/cosmos/ibc-go/v7 => ../../../
+toolchain go1.21.0
+
+replace github.com/cosmos/ibc-go/v8 => ../../../
 
 replace github.com/syndtr/goleveldb => github.com/syndtr/goleveldb v1.0.1-0.20210819022825-2ae1ddf74ef7
 
+replace github.com/cosmos/ibc-go/modules/light-clients/08-wasm => ../../../modules/light-clients/08-wasm
```
Comment (colin-axner) [inline modules/apps/callbacks/go.mod:11]: ditto

File: modules/light-clients/08-wasm/module.go [hunk 4]
```diff
@@ -114,7 +120,11 @@ func (AppModule) ProposalMsgs(simState module.SimulationState) []simtypes.Weight
 	return simulation.ProposalMsgs()
 }
 
-func (am AppModule) InitGenesis(ctx sdk.Context, cdc codec.JSONCodec, bz json.RawMessage) []abci.ValidatorUpdate {
+func (AppModule) ValidateGenesis(codec.JSONCodec, client.TxEncodingConfig, json.RawMessage) error {
+	return nil
+}
```
Comment (colin-axner) [inline modules/light-clients/08-wasm/module.go:125]: this comes from HasGenesis interface?

File: modules/light-clients/08-wasm/module.go [hunk 5]
```diff
@@ -133,10 +142,11 @@ func (am AppModule) ExportGenesis(ctx sdk.Context, cdc codec.JSONCodec) json.Raw
 }
 
 // BeginBlock implements the AppModule interface
-func (AppModule) BeginBlock(_ sdk.Context, _ abci.RequestBeginBlock) {
+func (AppModule) BeginBlock(context.Context) error {
+	return nil
 }
```
Comment (colin-axner) [inline modules/light-clients/08-wasm/module.go:147]: can remove

File: modules/light-clients/08-wasm/module.go [hunk 6]
```diff
@@ -133,10 +142,11 @@ func (am AppModule) ExportGenesis(ctx sdk.Context, cdc codec.JSONCodec) json.Raw
...
 
 // EndBlock implements the AppModule interface
-func (AppModule) EndBlock(_ sdk.Context, _ abci.RequestEndBlock) []abci.ValidatorUpdate {
-	return []abci.ValidatorUpdate{}
+func (AppModule) EndBlock(context.Context) error {
+	return nil
 }
```
Comment (colin-axner) [inline modules/light-clients/08-wasm/module.go:152]: can remove

**Review by colin-axner — APPROVED** [review]
LGTM! I actually reviewed all the relevant diffs by reviewing this pr in comparison to this branch against main. Happy to have any comments done in followups!

**Comment by chatton** [comment]
> Now that the wasm dep has been removed from ibc-go go.mod, the callbacks replace can be removed?

yep we should actually be able to remove a bunch of the hacks now 🥳

File: modules/light-clients/08-wasm/module.go (hunk 4 above)
Comment (chatton) [reply modules/light-clients/08-wasm/module.go:125]: that's right, from the embedded `HasGenesisBasics` interface

File: modules/apps/callbacks/Dockerfile (hunk 2 above)
Comment (chatton) [reply modules/apps/callbacks/Dockerfile:11]: should actually be able to remove now 🤞

**Comment by chatton** [comment]
@colin-axner I was able to remove all the hacks in the last commit 🥳

File: modules/light-clients/08-wasm/module.go [hunk 7]
```diff
@@ -18,18 +18,22 @@ import (
...
-	_ appmodule.AppModule    = (*AppModule)(nil)
+	_ module.AppModule           = (*AppModule)(nil)
+	_ module.AppModuleBasic      = (*AppModule)(nil)
+	_ module.HasProposalMsgs     = (*AppModule)(nil)
+	_ appmodule.AppModule        = (*AppModule)(nil)
+	_ module.HasGenesis          = (*AppModule)(nil)
+	_ module.HasName             = (*AppModule)(nil)
+	_ module.HasConsensusVersion = (*AppModule)(nil)
+	_ module.HasServices         = (*AppModule)(nil)
+	_ module.HasProposalMsgs     = (*AppModule)(nil)
```
Comment (charleenfei) [inline modules/light-clients/08-wasm/module.go:35]: i think these two lines might be duplicated!

**Review by charleenfei — APPROVED** [review]
one small comment but otherwise LGTM, thank you so much for this 🐰 🕳️

File: modules/light-clients/08-wasm/module.go (hunk 7 above)
Comment (chatton) [reply modules/light-clients/08-wasm/module.go:35]: ah nice catch

**Review by damiannolan — APPROVED** [review]

**Pull request merged**